be/src/vec/runtime/vdata_stream_recvr.h

```diff
@@ -219,23 +218,21 @@ class VDataStreamRecvr::PipSenderQueue : public SenderQueue {
     PipSenderQueue(VDataStreamRecvr* parent_recvr, int num_senders, RuntimeProfile* profile)
             : SenderQueue(parent_recvr, num_senders, profile) {}
 
-    bool should_wait() override {
-        return !_is_cancelled && _block_queue_empty && _num_remaining_senders > 0;
-    }
-
-    void _update_block_queue_empty() override { _block_queue_empty = _block_queue.empty(); }
-
     Status get_batch(Block* block, bool* eos) override {
+        std::lock_guard<std::mutex> l(_lock); // protect _block_queue
```

Comment:
this will make deadlock since we would lock `_lock` in `should_wait()`. maybe rewrite `should_wait` to `should_wait_without_lock`?